doc/source/reference/c-api/data_memory.rst

```diff
@@ -79,19 +79,17 @@ reallocate or free the data memory of the instance.
 
 .. c:function:: const PyDataMem_Handler * PyDataMem_SetHandler(PyDataMem_Handler *handler)
 
-   Sets a new allocation policy. If the input value is ``NULL``, will reset
-   the policy to the default. Returns the previous policy, ``NULL`` if the
-   previous policy was the default. We wrap the user-provided functions
+   Set a new allocation policy. If the input value is ``NULL``, will reset the
+   policy to the default. Return the previous policy, or set an exception and
+   return ``NULL`` if an error has occurred. We wrap the user-provided functions
    so they will still call the python and numpy memory management callback
    hooks.
     
-.. c:function:: const char * PyDataMem_GetHandlerName(PyArrayObject *obj)
+.. c:function:: const PyDataMem_Handler * PyDataMem_GetHandler(PyArrayObject *obj)
 
-   Return the const char name of the `PyDataMem_Handler` used by the
-   ``PyArrayObject`` or its base. If neither the ``PyArrayObject`` owns its own
-   data nor its base is a ``PyArrayObject`` which owns its own data return an
-   empty string. If ``NULL``, return the name of the current global policy
-   that will be used to allocate data for the next ``PyArrayObject``.
+   Return the `PyDataMem_Handler` used by the ``PyArrayObject``. If ``NULL``,
+   return the current global policy that will be used to allocate data for the
+   next ``PyArrayObject``. On failure, set an exception and return ``NULL``.
```

Comment:
```suggestion
   next ``PyArrayObject``. On failure, return ``NULL``.
```

Reply:
Most of the involved calls in these 2 functions (e.g. `PyCapsule_New`) set an exception (a `PyErr`) on error.

Your suggestion is not to document that our functions`set an exception`, or also to clear any exception (`PyErr_Clear`) before returning?

Reply:
When I look at the implementation of `PyDataMem_GetHandler`, I do not see any code that is setting an exception, so that is not something this function is doing. Whether an exception is set or not is the realm of  `PyContextVar_Get`, `PyCapsule_GetPointer`, and `PyArray_HANDLER`.

Reply:
What you are saying is true. Indeed claiming that on failure an exception is set is incorrect, since there are cases that this doesn't happen.

On the other hand, if our functions used functions that always `set an exception` on failure, I think that the natural way would be to document it.

I agree that my description is misleading, I will commit your suggestions. Thanks.